horovod/run/run_func.py

```diff
@@ -0,0 +1,199 @@
+                If None, use `os.environ`.
+    :param verbose: If this flag is set, extra messages will be printed.
+
+    :return: A list contains returned result from each MPI process. The ith value of the list
+             corresponds to the returned result of rank i process.
+    """
+
+    wdir = tempfile.mkdtemp()
+    if verbose:
+        print("horovod.run_func working dir is " + wdir)
+
+    if ssh_port:
+        ssh_port_opt = "-p " + str(ssh_port)
+        scp_port_opt = "-P " + str(ssh_port)
+    else:
+        ssh_port_opt = ""
+        scp_port_opt = ""
+
+    def wrapped_proc_fn(rank=0):
+        return_value = fn()
+
+        result_file = "{prefix}.{rank}".format(prefix=_PICKLED_RESULT_PREFIX, rank=rank)
+        with open(result_file, 'wb') as fp:
+            try:
+                cloudpickle.dump(return_value, fp)
+            except Exception as e:
+                raise RuntimeError("Caught an excpetion while pickling "
+                                   "return value: {}".format(type(e)), e)
+
+    pickled_proc_fn_str = cloudpickle.dumps(wrapped_proc_fn)
+    local_pickled_proc_fn_path = os.path.join(wdir, _LOCAL_PICKLED_PROC_FN_FILENAME)
+    with open(local_pickled_proc_fn_path, 'wb') as f:
+        f.write(pickled_proc_fn_str)
+
+    local_launcher_path = os.path.join(wdir, _LOCAL_PROC_LAUNCHER_FILENAME)
+    with open(local_launcher_path, 'w') as f:
+        f.write(textwrap.dedent("""
+        set -e
+        cd {wdir}
+        rank=${{OMPI_COMM_WORLD_RANK:-0}}
+        {exec_path} -c "import cloudpickle; cloudpickle.load(open('{fn_file}', 'rb'))(rank=$rank)"
+        """.format(wdir=wdir,
+                   exec_path=sys.executable,
+                   fn_file=_PICKLED_PROC_FN_FILENAME)))
+
+    if host is None:
+        host = "localhost:{np}".format(np=num_proc)
+
+    all_host_names = set(x.split(':')[0] for x in host.split(','))
+
+    # copy pickled function and launcher script to all hosts via scp
+    # TODO: parallelize the scp copy
+    for hostname in all_host_names:
+        scp_cmd = textwrap.dedent(
+            """
+            ssh {ssh_port_opt} -o StrictHostKeyChecking=no {hostname} mkdir -p {wdir} && \
+            scp -r -q {scp_port_opt} -o StrictHostKeyChecking=no \
+            {wdir}/{local_fn_file} {hostname}:{wdir}/{fn_file} && \
+            scp -r -q {scp_port_opt} -o StrictHostKeyChecking=no \
+            {wdir}/{local_launcher} {hostname}:{wdir}/{launcher}
+            """).format(ssh_port_opt=ssh_port_opt, scp_port_opt=scp_port_opt,
+                        wdir=wdir, hostname=hostname,
+                        local_fn_file=_LOCAL_PICKLED_PROC_FN_FILENAME,
+                        fn_file=_PICKLED_PROC_FN_FILENAME,
+                        local_launcher=_LOCAL_PROC_LAUNCHER_FILENAME,
+                        launcher=_PROC_LAUNCHER_FILENAME)
+
+        output = six.StringIO()
+        exit_code = safe_shell_exec.execute(scp_cmd, stdout=output, stderr=output)
```

Comment:
I read `safe_shell_exec` code. It is a wrapper for `subprocess.Popen`. But it add two features:
* when parent process exit, force child process exit.
* forward child process stdout/stderr to user-specific stream. We can set it to redirect to parent process stdout/stderr.

Btw, I add a arg `disable_stdin` for `safe_shell_exec`.

So that I suggest to use `safe_shell_exec` in every places.